### PR TITLE
Feat/playback - Improve and Refactor playback profiles

### PR DESCRIPTION
**Pull request by Camc314**
This should:

 - Do not use AC3 for audio transcoding if AAC and MP3 are supported (based on jellyfin/jellyfin-web#2135)
 - This should enable HEVC over FMP4-HLS (based on jellyfin/jellyfin-web#2064)
 - Adjust the default audio codec to AAC for HLS streaming (based on jellyfin/jellyfin-web#2026)

This also moves playback profile files to more relevant folders

### CODE REVIEW

File: utils/playbackProfiles/helpers/codecProfiles.ts [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+      console.error('isUdPanelSupported() error code = ' + error.code);
+    }
+  }
+
+  return browserDetector.isPs4()
+    ? 8000000
+    : browserDetector.isXbox()
+    ? 12000000
+    : browserDetector.isTizen && isTizenFhd
+    ? 20000000
```
Comment (ferferga) [inline utils/playbackProfiles/helpers/codecProfiles.ts:38]: I think we should document this kind of hardcoded values better

File: utils/playbackProfiles/helpers/codecProfiles.ts [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+    Condition,
+    Property,
+    Value,
+    IsRequired
+  };
+}
+
+/**
+ * @param {HTMLVideoElement} videoTestElement A HTML video element for testing codecs
+ * @returns {ProfileCondition[]} sd
```
Comment (ferferga) [inline utils/playbackProfiles/helpers/codecProfiles.ts:67]: I think you missed changing this?

File: utils/playbackProfiles/helpers/codecProfiles.ts [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+
+/**
+ *
+ *
+ * @param {HTMLVideoElement} videoTestElement A HTML video element for testing codecs
+ * @returns {Array<CodecProfile>} - Array containing the different profiles for the client
+ */
+export function getCodecProfiles(
+  videoTestElement: HTMLVideoElement
+): Array<CodecProfile> {
```
Comment (ferferga) [inline utils/playbackProfiles/helpers/codecProfiles.ts:110]:
```suggestion
): CodecProfile[] {
```

File: utils/playbackProfiles/helpers/codecProfiles.ts [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+    maxH264Level = 52;
+  }
+
+  if (
+    browserDetector.isTizen() ||
+    videoTestElement
+      .canPlayType('video/mp4; codecs="avc1.6e0033"')
+      .replace(/no/, '')
+  ) {
+    // These tests are passing in safari, but playback is failing
```
Comment (ferferga) [inline utils/playbackProfiles/helpers/codecProfiles.ts:163]:
```suggestion
    // TODO: These tests are passing in safari, but playback is failing
```

File: utils/playbackProfiles/helpers/codecProfiles.ts [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+    });
+  }
+
+  return conditions;
+}
+
+/**
+ *
+ *
+ * @param {HTMLVideoElement} videoTestElement A HTML video element for testing codecs
```
Comment (ferferga) [inline utils/playbackProfiles/helpers/codecProfiles.ts:105]:
```suggestion
 * @param {HTMLVideoElement} videoTestElement - A HTML video element for testing codecs
```

There are more JSDoc in the PR that needs this.

File: utils/playbackProfiles/helpers/fmp4AudioFormats.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  hasAacSupport,
+  hasAc3InHlsSupport,
+  hasAc3Support,
+  hasEac3Support,
+  hasMp3AudioSupport
+} from './mp4AudioFormats';
+
+/**
+ * @param {HTMLVideoElement} videoTestElement A HTML video element for testing codecs
+ * @returns {string[]} List of ...
```
Comment (ferferga) [inline utils/playbackProfiles/helpers/fmp4AudioFormats.ts:12]:
```suggestion
 * @param {HTMLVideoElement} videoTestElement - A HTML video element for testing codecs
 * @returns {string[]} List of supported FMP4 audio codecs
```

File: utils/playbackProfiles/helpers/mp4AudioFormats.ts [hunk 7]
```diff
@@ -107,7 +161,7 @@ export function getSupportedMP4AudioCodecs(
     codecs.push('dts');
   }
 
-  if (browserDetector.isTv()) {
+  if (browserDetector.isTizen() || browserDetector.isWebOS()) {
```
Comment (ferferga) [inline utils/playbackProfiles/helpers/mp4AudioFormats.ts:164]: What about TVs with neither, are we planning improvements on that side?

Comment (MrTimscampi) [reply utils/playbackProfiles/helpers/mp4AudioFormats.ts:164]: There aren't really any actively supported smart TV OSes aside from these two and Android TV. All the others are unsupported.

File: utils/playbackProfiles/helpers/codecProfiles.ts (hunk 2 above)
Comment (Camc314) [reply utils/playbackProfiles/helpers/codecProfiles.ts]: Yep. Should be updated

File: utils/playbackProfiles/helpers/codecProfiles.ts (hunk 1 above)
Comment (Camc314) [reply utils/playbackProfiles/helpers/codecProfiles.ts]: I have just copied these values from Jf-web. I'm not entirely sure where they come from. I can add a todo to investigate where they come from?

File: utils/playbackProfiles/helpers/codecProfiles.ts (hunk 5 above)
Comment (Camc314) [reply utils/playbackProfiles/helpers/codecProfiles.ts:102]: Probably easiest to do them all in a single PR. I'll make one after this is merged

File: utils/playbackProfiles/helpers/codecProfiles.ts (hunk 1 above)
Comment (ferferga) [reply utils/playbackProfiles/helpers/codecProfiles.ts]: This is exactly why I mentioned this. For now I think it's enough, nothing else we can do then.

**Review by ferferga — APPROVED** [review]
Once the PS4 comment thing is addressed, this is good to go

File: utils/playbackProfiles/helpers/codecProfiles.ts (hunk 1 above)
Comment (Camc314) [reply utils/playbackProfiles/helpers/codecProfiles.ts]: Cool, A comment has been added.

**Comment by Camc314** [comment]
> Once the PS4 comment thing is addressed, this is good to go

Comment Added

**Pull request merged**